src/omv/imlib/draw.c

```diff
@@ -3061,19 +3020,50 @@ void imlib_draw_image(image_t *dst_img,
     size_t dst_img_row_bytes = image_size(dst_img) / dst_img->h;
 
     // Do we need to convert the image?
-    bool is_bayer_color_conversion = src_img->is_bayer && !dst_img->is_bayer;
-    bool is_yuv_color_conversion = src_img->is_yuv && !dst_img->is_yuv;
-    bool is_color_conversion = is_bayer_color_conversion || is_yuv_color_conversion;
+    bool is_bayer_conversion = src_img->is_bayer && !dst_img->is_bayer;
+    bool is_yuv_conversion = src_img->is_yuv && !dst_img->is_yuv;
+    bool is_bayer_yuv_conversion = is_bayer_conversion || is_yuv_conversion;
 
-    // Force a deep copy if we cannot use the image in-place.
-    bool need_deep_copy = (dst_img->data == src_img->data)
-                          && (is_scaling || (src_img_row_bytes < dst_img_row_bytes) || is_color_conversion);
+    // Is the line length growing which will prevent us from working in-place?
+    bool is_line_length_growing = src_img_row_bytes < dst_img_row_bytes;
 
-    // Force a deep copy if we are scaling.
-    bool is_color_conversion_scaling = is_color_conversion && is_scaling;
+    #if (OMV_GPU_ENABLE == 1)
+    if ((!callback) &&
+        (!is_bayer_yuv_conversion) &&
+        (!src_img->is_compressed) &&
+        (rgb_channel < 0) &&
+        ((dst_img->data != src_img->data) || (!is_line_length_growing)) &&
+        (!(hint & (IMAGE_HINT_AREA | IMAGE_HINT_BICUBIC | IMAGE_HINT_TRANSPOSE)))) {
+
+        rectangle_t dst_rect = {
+            .x = dst_x_start,
+            .y = dst_y_start,
+            .w = dst_x_end - dst_rect.x,
+            .h = dst_y_end - dst_rect.y,
+        };
+
+        rectangle_t src_rect = {
+            .x = src_x_accum_reset >> 16,
+            .y = src_y_accum_reset >> 16,
+            .w = ((src_x_accum_reset + (src_x_frac * dst_rect.w)) >> 16) - src_rect.x,
+            .h = ((src_y_accum_reset + (src_y_frac * dst_rect.h)) >> 16) - src_rect.y,
+        };
+
+        image_hint_t gpu_hints = ((dst_delta_x < 0) ? IMAGE_HINT_HMIRROR : 0) |
+                                 ((dst_delta_y < 0) ? IMAGE_HINT_VFLIP : 0) |
+                                 (hint & (IMAGE_HINT_BILINEAR | IMAGE_HINT_BLACK_BACKGROUND));
+
+        if (!omv_gpu_draw_image(src_img, &src_rect, dst_img, &dst_rect,
+                                alpha, color_palette, alpha_palette, gpu_hints)) {
+            goto exit_cleanup;
+        }
+    }
+    #endif
 
     // Make a deep copy of the source image.
-    if (need_deep_copy || is_color_conversion_scaling || is_jpeg || is_png) {
+    if (((dst_img->data == src_img->data) && (is_scaling || is_line_length_growing || is_bayer_yuv_conversion))
+        || (is_bayer_yuv_conversion && is_scaling)
+        || src_img->is_compressed) {
```

Comment:
```suggestion
    if (((dst_img->data == src_img->data) &&
        (is_scaling || is_line_length_growing || is_bayer_yuv_conversion)) ||
        (is_bayer_yuv_conversion && is_scaling) ||
        src_img->is_compressed) {
```

Reply:
Done

Reply:
```
if (((dst_img->data == src_img->data) &&
         (is_scaling || is_upscaling || is_bayer_yuv_conversion)) ||
        (is_bayer_yuv_conversion && is_scaling) ||
        src_img->is_compressed) {
```

Note: Clang requires a space on the `(is_scaling || is_upscaling || is_bayer_yuv_conversion))` line since it's part of the compound statement of the first line.

Reply:
Sure, the point is the logical ops go on the end.